Reining on over eager optimizations (take 2)

retronym opened the pull request:
This time I allow elimination of dead, non-static loads
of closure classes.

Review by @lrytz @dragos @gkossakowski

retronym committed:
SI-8601 Don't treat int/long division, or arraylength, as dead-code

`{i, l}div` and `{i, l}rem` throw an `ArithmeticException` if the
divisor is 0.

`arraylength` throws a `NullPointerException` on a null reference.

JVM Spec:
> The only integer operations that can throw an exception are the
> integer divide instructions (idiv and ldiv) and the integer
> remainder instructions (irem and lrem), which throw an
> ArithmeticException if the divisor is zero.

> The Java virtual machine's floating-point operators do not throw
> runtime exceptions

> If the arrayref is null, the arraylength instruction throws a
> NullPointerException.

I checked the other primitives in `ICode` to see if anything else
should be considered as live code.

Pure:

	  // jvm  : {i, l, f, d}neg
	  case class Negation(kind: TypeKind) extends Primitive
	  // jvm  : if{eq, ne, lt, ge, le, gt}, if{null, nonnull}
	  //        if_icmp{eq, ne, lt, ge, le, gt}, if_acmp{eq,ne}
	  case class Test(op: TestOp, kind: TypeKind,  zero: Boolean)  extends Primitive
	  // jvm  : lcmp, {f, d}cmp{l, g}
	  case class Comparison(op: ComparisonOp, kind: TypeKind) extends Primitive

Impure: {i, l}{div, rem}, otherwise pure

	  // jvm  : {i, l, f, d}{add, sub, mul, div, rem}
	  case class Arithmetic(op: ArithmeticOp, kind: TypeKind) extends Primitive

Pure (overflow is silent, NaN.toInt is defined):
	  // jvm  : {i, l}{and, or, xor}
	  case class Logical(op: LogicalOp, kind: TypeKind) extends Primitive

	  // jvm  : {i, l}{shl, ushl, shr}
	  case class Shift(op: ShiftOp, kind: TypeKind) extends Primitive

	  // jvm  : i2{l, f, d}, l2{i, f, d}, f2{i, l, d}, d2{i, l, f}, i2{b, c, s}
	  case class Conversion(src: TypeKind, dst: TypeKind) extends Primitive

Impure! May NPE!

	  // jvm  : arraylength
	  case class ArrayLength(kind: TypeKind) extends Primitive

Pure (we know that StringBuilder.{<init>, append, toString} are pure
and `append` is null safe.)

	  // jvm  : It should call the appropiate 'append' method on StringBuffer
	  case class StringConcat(el: TypeKind) extends Primitive

	  // jvm:  it should create a new StringBuffer
	  case object StartConcat extends Primitive

      // jvm:  convert StringBuffer to a String
	  case object EndConcat extends Primitive

retronym committed:
SI-8601 Avoid over-eager optimization of LOAD_FIELD

It can NPE or trigger static class initilization, we can't elimiate
it without changing semantics.

To make sure we don't thwart closure elimination, I've allowed DCE
to eliminate a non-static LOAD_FIELD of a member of a closure class.
It would be more general to track nullity of the reciever (e.g, `this`
or `new Foo` cannot be null), but that would require more
infrastructure in this phase.

I've added a test for closure inlining based on a a suggestion by
@dragos. This actually passes if we remove the (LOAD_FIELD, DROP)
peephole optimization for `closelim` altogether. But I chose to
adapt that optimization (only allow it for non-static, closure
fields), rather then remove it alogether, in the interests of
treading lightly.

retronym committed:
Add a test for array load

retronym committed:
SI-8601 Don't treat newarray as dead code

Otherwise we lose the side effect of a `NegativeArraySizeException`.

retronym committed:
SI-8601 Test that `null.synchronized` NPEs under -optimize

As part of my sweep through the side-effecting byte code instructions.

retronym commented:
Tweaked resubmission of #3760. ccc5eef is the only different commit.

retronym commented:
I'll add a few more tests tomorrow for static initialiser side effects to make sure the isstatic logic is correct.

lrytz commented:
LGTM. The new backend supports dataflow analysis to [track nullness / non-nullness](https://github.com/lrytz/scala/blob/opt/rebase/src/compiler/scala/tools/nsc/backend/bcode/NullnessPropagator.java), which I guess can be used for that eliminating dead loads.

retronym commented:
Let's also check that the size of the bootstrapped compiler hasn't changed before merging this. I'm going to close the PR pending that, and pending my extra tests, so we don't merge it yet.

retronym commented:
BTW, `-Yconst-opt` also tracks nullity in the old backend.

gkossakowski commented on test/files/run/t8601-closure-elim.scala:
```diff
@@ -0,0 +1,26 @@
+      val classNode = loadClassNode("Foo")
+      val methodNode = getMethod(classNode, "b")
+      val ops = methodNode.instructions.iterator.asScala.map(_.getOpcode).toList
+      assert(!ops.contains(asm.Opcodes.NEW), ops)// should be allocation free if the closure is eliminiated
```
Have you considered writing the test like [inline-in-constructors](https://github.com/scala/scala/tree/2.12.x/test/files/instrumented/inline-in-constructors) ?

retronym replied:
How do you see the pros and cons of each approach?

gkossakowski replied:
Pros:
- instrumentation let's you asses runtime behavior of a program which is what we are after in most cases when doing functional testing of the optimizer
- since you are counting method invocations you are less concerned with details of bytecode generation (I can imagine that you want to have some `new` instruction in your method but you want to know that no constructor of anonfun is called)

Cons:
- relies on bytecode instrumentation

I think in your specific case both ways will do. I wanted to point out at existing test for the optimizer, covering similar area.

retronym commented:
```
% jar tf build/pack/lib/scala-compiler.jar | sort > sandbox/pack.txtticket/8601
% jar tf $(scala-hash 2.11.x -h)/lib/scala-compiler.jar | sort > sandbox/2.11.x.txt
[warn] 2.11.x failed, using closest available

% diff -u sandbox/{2.11.x,pack}.txt | subl
```

``` patch
--- sandbox/2.11.x.txt  2014-05-20 11:00:07.000000000 +0200
+++ sandbox/pack.txt    2014-05-20 10:59:52.000000000 +0200
@@ -1945,6 +1945,8 @@
 scala/tools/nsc/backend/icode/analysis/TypeFlowAnalysis$typeStackLattice$.class
 scala/tools/nsc/backend/icode/analysis/TypeFlowAnalysis.class
 scala/tools/nsc/backend/jvm/
+scala/tools/nsc/backend/jvm/AsmUtils$.class
+scala/tools/nsc/backend/jvm/AsmUtils.class
 scala/tools/nsc/backend/jvm/BCodeBodyBuilder$PlainBodyBuilder$$anonfun$1.class
 scala/tools/nsc/backend/jvm/BCodeBodyBuilder$PlainBodyBuilder$$anonfun$2.class
 scala/tools/nsc/backend/jvm/BCodeBodyBuilder$PlainBodyBuilder$$anonfun$4.class
@@ -4733,6 +4735,7 @@
 scala/tools/nsc/plugins/Plugins$$anonfun$6.class
 scala/tools/nsc/plugins/Plugins$$anonfun$7.class
 scala/tools/nsc/plugins/Plugins$$anonfun$8.class
+scala/tools/nsc/plugins/Plugins$$anonfun$computePluginPhases$1$$anonfun$apply$6.class
 scala/tools/nsc/plugins/Plugins$$anonfun$computePluginPhases$1.class
 scala/tools/nsc/plugins/Plugins$$anonfun$loadPlugins$1$$anonfun$apply$3.class
 scala/tools/nsc/plugins/Plugins$$anonfun$loadPlugins$1.class
@@ -6012,7 +6015,6 @@
 scala/tools/nsc/transform/patmat/MatchCodeGen$CodegenCore.class
 scala/tools/nsc/transform/patmat/MatchCodeGen$OptimizedCodegen$class.class
 scala/tools/nsc/transform/patmat/MatchCodeGen$OptimizedCodegen$optimizedCodegen$$anonfun$1.class
-scala/tools/nsc/transform/patmat/MatchCodeGen$OptimizedCodegen$optimizedCodegen$$anonfun$2$$anonfun$3.class
 scala/tools/nsc/transform/patmat/MatchCodeGen$OptimizedCodegen$optimizedCodegen$$anonfun$2.class
 scala/tools/nsc/transform/patmat/MatchCodeGen$OptimizedCodegen$optimizedCodegen$.class
 scala/tools/nsc/transform/patmat/MatchCodeGen$OptimizedCodegen$optimizedCodegen$OptimizedCasegen$$anonfun$7.class
@@ -6056,7 +6058,6 @@
 scala/tools/nsc/transform/patmat/MatchOptimization$CommonSubconditionElimination$ReusedCondTreeMaker$$anonfun$treesToHoist$1.class
 scala/tools/nsc/transform/patmat/MatchOptimization$CommonSubconditionElimination$ReusedCondTreeMaker$.class
 scala/tools/nsc/transform/patmat/MatchOptimization$CommonSubconditionElimination$ReusedCondTreeMaker.class
-scala/tools/nsc/transform/patmat/MatchOptimization$CommonSubconditionElimination$ReusingCondTreeMaker$$anonfun$12$$anonfun$apply$13.class
 scala/tools/nsc/transform/patmat/MatchOptimization$CommonSubconditionElimination$ReusingCondTreeMaker$$anonfun$12$$anonfun$apply$14.class
 scala/tools/nsc/transform/patmat/MatchOptimization$CommonSubconditionElimination$ReusingCondTreeMaker$$anonfun$12.class
 scala/tools/nsc/transform/patmat/MatchOptimization$CommonSubconditionElimination$ReusingCondTreeMaker$$anonfun$13.class
@@ -6960,7 +6961,6 @@
 scala/tools/nsc/typechecker/Namers$Namer$$anonfun$addDefaultGetters$2$$anonfun$apply$14$$anonfun$14.class
 scala/tools/nsc/typechecker/Namers$Namer$$anonfun$addDefaultGetters$2$$anonfun$apply$14$$anonfun$15.class
 scala/tools/nsc/typechecker/Namers$Namer$$anonfun$addDefaultGetters$2$$anonfun$apply$14$$anonfun$16$$anonfun$17.class
-scala/tools/nsc/typechecker/Namers$Namer$$anonfun$addDefaultGetters$2$$anonfun$apply$14$$anonfun$16$$anonfun$apply$16.class
 scala/tools/nsc/typechecker/Namers$Namer$$anonfun$addDefaultGetters$2$$anonfun$apply$14$$anonfun$16.class
 scala/tools/nsc/typechecker/Namers$Namer$$anonfun$addDefaultGetters$2$$anonfun$apply$14$$anonfun$18.class
 scala/tools/nsc/typechecker/Namers$Namer$$anonfun$addDefaultGetters$2$$anonfun$apply$14$$anonfun$19.class
@@ -7195,6 +7195,7 @@
 scala/tools/nsc/typechecker/RefChecks$RefCheckTransformer$$anonfun$transform$3.class
 scala/tools/nsc/typechecker/RefChecks$RefCheckTransformer$$anonfun$transformCaseApply$1.class
 scala/tools/nsc/typechecker/RefChecks$RefCheckTransformer$$anonfun$transformSelect$2.class
+scala/tools/nsc/typechecker/RefChecks$RefCheckTransformer$$anonfun$transformStat$1.class
 scala/tools/nsc/typechecker/RefChecks$RefCheckTransformer$$anonfun$transformStat$2.class
 scala/tools/nsc/typechecker/RefChecks$RefCheckTransformer$$anonfun$transformStat$3.class
 scala/tools/nsc/typechecker/RefChecks$RefCheckTransformer$$anonfun$transformStats$1.class
@@ -7632,6 +7633,7 @@
 scala/tools/nsc/typechecker/Typers$Typer$$anonfun$packCaptured$1.class
 scala/tools/nsc/typechecker/Typers$Typer$$anonfun$packedTypes$1.class
 scala/tools/nsc/typechecker/Typers$Typer$$anonfun$reenterTypeParams$1.class
+scala/tools/nsc/typechecker/Typers$Typer$$anonfun$reenterValueParams$1$$anonfun$apply$5.class
 scala/tools/nsc/typechecker/Typers$Typer$$anonfun$reenterValueParams$1.class
 scala/tools/nsc/typechecker/Typers$Typer$$anonfun$rewrappingWrapperTrees$1$$anonfun$apply$15.class
 scala/tools/nsc/typechecker/Typers$Typer$$anonfun$rewrappingWrapperTrees$1$$anonfun$apply$16.class
```

retronym committed:
SI-8601 Test that static LOAD_FIELD is not eliminated

This test fails under 2.11.0, and works now that DCE treats
static loads as useful instructions.

retronym commented:
@lrytz I've added the test for static init side effects.

lrytz commented:
LGTM
